packages/gatsby/src/internal-plugins/functions/gatsby-node.ts

```diff
@@ -41,42 +110,51 @@ const createWebpackConfig = async ({
     `functions`
   )
 
-  const files = await new Promise((resolve, reject) => {
-    glob(`**/*.{js,ts}`, { cwd: functionsDirectory }, (err, files) => {
-      if (err) {
-        reject(err)
-      } else {
-        resolve(files)
-      }
-    })
-  })
+  const globs = createGlobArray(
+    siteDirectoryPath,
+    store.getState().flattenedPlugins
+  )
 
-  if (files?.length === 0) {
-    reporter.warn(
-      `No functions found in directory: ${path.relative(
-        siteDirectoryPath,
-        functionsDirectory
-      )}`
-    )
-  }
+  // Glob and return object with relative/absolute paths + which plugin
+  // they belong to.
+  const allFunctions = await Promise.all(
+    globs.map(async glob => {
+      const knownFunctions: Array<IFunctionData> = []
+      const files = await globAsync(glob.globPattern)
+      files.map(file => {
+        const originalAbsoluteFilePath = file
+        const originalRelativeFilePath = path.relative(glob.rootPath, file)
+
+        const { dir, name } = path.parse(originalRelativeFilePath)
+        // Ignore the original extension as all compiled functions now end with js.
+        const compiledFunctionName = path.join(dir, name + `.js`)
+        const compiledPath = path.join(
+          compiledFunctionsDir,
+          compiledFunctionName
+        )
+        const finalName = urlResolve(dir, name === `index` ? `` : name)
+
+        knownFunctions.push({
+          functionRoute: finalName,
+          pluginName: glob.pluginName,
+          originalAbsoluteFilePath,
+          originalRelativeFilePath,
+          relativeCompiledFilePath: compiledFunctionName,
+          absoluteCompiledFilePath: compiledPath,
+          matchPath: getMatchPath(finalName),
+        })
+      })
 
-  const knownFunctions: Array<IFunctionData> = []
-  knownFunctions.forEach(f => f.apiRoute)
-  files.map(file => {
-    const { dir, name } = path.parse(file)
-    // Ignore the original extension as all compiled functions now end with js.
-    const compiledFunctionName = path.join(dir, name + `.js`)
-    const compiledPath = path.join(compiledFunctionsDir, compiledFunctionName)
-    const finalName = urlResolve(dir, name === `index` ? `` : name)
-
-    knownFunctions.push({
-      apiRoute: finalName,
-      originalFilePath: file,
-      relativeCompiledFilePath: compiledFunctionName,
-      absoluteCompiledFilePath: compiledPath,
-      matchPath: getMatchPath(finalName),
+      return knownFunctions
     })
-  })
+  )
+
+  // Combine functions by the route name so that functions in the default
+  // functions directory can override the plugin's implementations.
+  const knownFunctions = _.unionBy(
```

Comment:
we implement shadow using lodash's `unionBy` and combine by the `functionRoute`. The site's function implementations are first so they override any matching functions that come later.